x-pack/plugins/apm/public/components/app/TransactionDetails/Distribution/index.tsx

```diff
@@ -47,191 +48,212 @@ export function getFormattedBuckets(buckets: IBucket[], bucketSize: number) {
-      'xpack.apm.transactionDetails.transactionsDurationDistributionChart.unitShortLabel',
-      {
-        defaultMessage:
-          '{transCount} {transType, select, request {req.} other {trans.}}',
-        values: {
-          transCount: t,
-          transType: this.props.urlParams.transactionType
-        }
-      }
-    );
-  };
-
-  public formatYLong = (t: number) => {
-    return this.props.urlParams.transactionType === 'request'
-      ? i18n.translate(
-          'xpack.apm.transactionDetails.transactionsDurationDistributionChart.requestTypeUnitLongLabel',
-          {
-            defaultMessage:
-              '{transCount, plural, =0 {# request} one {# request} other {# requests}}',
-            values: {
-              transCount: t
-            }
-          }
-        )
-      : i18n.translate(
-          'xpack.apm.transactionDetails.transactionsDurationDistributionChart.transactionTypeUnitLongLabel',
-          {
-            defaultMessage:
-              '{transCount, plural, =0 {# transaction} one {# transaction} other {# transactions}}',
-            values: {
-              transCount: t
-            }
-          }
-        );
-  };
+export const TransactionDistribution: FunctionComponent<Props> = (
+  props: Props
+) => {
+  const {
+    location: { pathname, search, hash },
+    distribution,
+    urlParams: { transactionId, traceId, transactionType }
+  } = props;
 
-  public redirectToTransactionType() {
-    const { urlParams, location, distribution } = this.props;
+  const formatYShort = useCallback(getFormatYShort(transactionType), [
+    transactionType
+  ]);
 
-    if (
-      !distribution ||
-      !distribution.defaultSample ||
-      urlParams.traceId ||
-      urlParams.transactionId
-    ) {
-      return;
-    }
+  const formatYLong = useCallback(getFormatYLong(transactionType), [
+    transactionType
+  ]);
 
-    const { traceId, transactionId } = distribution.defaultSample;
+  const redirectToDefaultSample = useCallback(
+    () => {
+      const defaultSample =
+        distribution && distribution.defaultSample
+          ? distribution.defaultSample
+          : null;
 
-    history.replace({
-      ...location,
-      search: fromQuery({
-        ...toQuery(location.search),
-        traceId,
-        transactionId
-      })
-    });
-  }
+      const parsedQueryParams = toQuery(search);
 
-  public componentDidMount() {
-    this.redirectToTransactionType();
-  }
+      const nextTransaction = defaultSample
+        ? {
+            traceId: defaultSample.traceId,
+            transactionId: defaultSample.transactionId
+          }
+        : {};
 
-  public componentDidUpdate() {
-    this.redirectToTransactionType();
-  }
+      history.replace({
+        pathname,
+        hash,
+        search: fromQuery({
+          ...omit(parsedQueryParams, 'transactionId', 'traceId'),
+          ...nextTransaction
+        })
+      });
+    },
+    [distribution, pathname, hash, search]
```

Comment:
This might be correct, but I'm wondering:
if we care about `pathname`, `hash` and `search`, why not pass in `[distribution, location]`?

On the other hand: do we even care about location changing? Isn't it only `[distribution, transactionId, traceId]` we care about?

Reply:
I'll check again, but I think this is because if we pass location as a dependency to `redirectToDefaultSample`, it's identity changes every time location changes (even though its _value_ has not changed. Because we pass the callback to the `useEffect` hook as a dependency, that hook is being executed too often as well. I might be able to clean it up though.

Reply:
Yeah, the renderer gets stuck in a loop. The initial render triggers a redirect to the default sample of the selected transaction distribution. Because of that redirect, the (identity of the) `location` object changes. That will trigger the redirect hook to be called on successive renders too, ad infinitum. However, because the values of the `location` object do not change, we can prevent this by using the primitives.

Reply:
I see. What do you think about getting location from the `history.location` singleton? `history.location` is a mutable object, so cannot be used to detect changes but that is also not needed here (on the contrary actually). So something like:

```js
  const redirectToDefaultSample = useCallback(
    () => {    
      const parsedQueryParams = toQuery(history.location.search);
      const defaultSample =  idx(distribution, _ => _.defaultSample) || {};

      history.replace({
        ...history.location,
        search: fromQuery({
          ...omit(parsedQueryParams, 'transactionId', 'traceId'),
          ...defaultSample
        })
      });
    },
    [distribution]
  );
```

Of course this only works if all we care about are changes to `distribution`.